merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/CheckpointSimulationDriver.java

```diff
@@ -0,0 +1,412 @@
+import gov.nasa.jpl.aerie.merlin.protocol.driver.Topic;
+import gov.nasa.jpl.aerie.merlin.protocol.model.Task;
+import gov.nasa.jpl.aerie.merlin.protocol.model.TaskFactory;
+import gov.nasa.jpl.aerie.merlin.protocol.types.Duration;
+import gov.nasa.jpl.aerie.merlin.protocol.types.InstantiationException;
+import org.apache.commons.lang3.mutable.MutableLong;
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static gov.nasa.jpl.aerie.merlin.protocol.types.Duration.MAX_VALUE;
+import static gov.nasa.jpl.aerie.merlin.protocol.types.Duration.ZERO;
+import static gov.nasa.jpl.aerie.merlin.protocol.types.Duration.min;
+
+public class CheckpointSimulationDriver {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CheckpointSimulationDriver.class);
+
+  public record CachedSimulationEngine(
+      Duration endsAt,
+      Map<ActivityDirectiveId, ActivityDirective> activityDirectives,
+      SimulationEngine simulationEngine,
+      LiveCells cells,
+      SlabList<TemporalEventSource.TimePoint> timePoints,
+      Topic<ActivityDirectiveId> activityTopic,
+      MissionModel<?> missionModel
+  ) {
+    public void freeze() {
+      cells.freeze();
+      timePoints.freeze();
+      simulationEngine.close();
+    }
+
+    public static CachedSimulationEngine empty(final MissionModel<?> missionModel) {
+      final SimulationEngine engine = new SimulationEngine();
+      final TemporalEventSource timeline = new TemporalEventSource();
+      final LiveCells cells = new LiveCells(timeline, missionModel.getInitialCells());
+
+      // Begin tracking all resources.
+      for (final var entry : missionModel.getResources().entrySet()) {
+        final var name = entry.getKey();
+        final var resource = entry.getValue();
+
+        engine.trackResource(name, resource, Duration.ZERO);
+      }
+
+      {
+        // Start daemon task(s) immediately, before anything else happens.
+        engine.scheduleTask(Duration.ZERO, missionModel.getDaemon());
+        {
+          final var batch = engine.extractNextJobs(Duration.MAX_VALUE);
+          final var commit = engine.performJobs(batch.jobs(), cells, Duration.ZERO, Duration.MAX_VALUE);
+          timeline.add(commit.getKey());
+        }
+      }
+      final var emptyCachedEngine = new CachedSimulationEngine(
+          Duration.MIN_VALUE,
+          Map.of(),
+          engine,
+          cells,
+          timeline.points(),
+          new Topic<>(),
+          missionModel
+      );
+      return emptyCachedEngine;
```

Comment:
Minor, nonblocking: Simplify:
```suggestion
      return new CachedSimulationEngine(
          Duration.MIN_VALUE,
          Map.of(),
          engine,
          cells,
          timeline.points(),
          new Topic<>(),
          missionModel
      );
```